### PR TITLE
Create internals for testing IPv6 network between lib and agent

### PR DESCRIPTION
**Pull request by cbeauchesne**
## Motivation

#3549

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

**Review by mhlidd — APPROVED** [review]
From the weblog testing side it looks good to me, although I would have someone else look at the container changes that were made.

File: .github/workflows/run-end-to-end.yml [hunk 1]
```diff
@@ -115,6 +116,9 @@ jobs:
     - name: Run DEFAULT scenario
       if: always() && steps.build.outcome == 'success' && contains(inputs.scenarios, '"DEFAULT"')
       run: ./run.sh DEFAULT
+    - name: Run IPV6 scenario
+      if: always() && steps.build.outcome == 'success' && contains(inputs.scenarios, '"IPV6"') && !contains(fromJSON('["cpp", "dotnet", "java", "ruby", "ipv6"]'), inputs.library)
```
Comment (robertomonteromiguel) [inline .github/workflows/run-end-to-end.yml:120]: Why ipv6 here?

`!contains(fromJSON('["cpp", "dotnet", "java", "ruby", "ipv6"]')`

File: docs/execute/IPv6.md [hunk 2]
```diff
@@ -0,0 +1 @@
+Some scenarios requires the docker daemon to supports IPv6. It should be ok on Linux and Mac OS, but has not be tested on Windows.
```
Comment (robertomonteromiguel) [inline docs/execute/IPv6.md:1]: is this file referenced in somewhere?

File: utils/_context/_scenarios/endtoend.py [hunk 3]
```diff
@@ -389,18 +417,35 @@ def _start_interface_watchdog(self):
...
+
+            if sys.platform == "linux":
+                # on Linux, with ipv6 mode, we can't use localhost anymore for a reason I ignore
+                # To fix, we use the container ipv4 address as weblog doamin, as it's accessible from host
+
+                weblog.domain = self.weblog_container.network_ip(self._network)
+            elif sys.platform == "darwin":
+                # on Mac, this ipv4 address can't be used, because container IP are not accessible from host
+                # as they are on an network intermal to the docker VM. But we can still use localhost.
+                ...
```
Comment (robertomonteromiguel) [inline utils/_context/_scenarios/endtoend.py:432]: add a logger.info?

File: utils/_context/containers.py [hunk 4]
```diff
@@ -181,12 +187,20 @@ def start(self, network: Network) -> Container:
         if self.healthy:
             self.warmup()
 
+        self._container.reload()
```
Comment (robertomonteromiguel) [inline utils/_context/containers.py:190]: why reload?

File: utils/_context/containers.py [hunk 5]
```diff
@@ -181,12 +187,20 @@ def start(self, network: Network) -> Container:
...
+            json.dump(self._container.attrs, f, indent=2)
+
     def async_start(self, network: Network) -> Thread:
         """Start the container and its dependencies in a thread with circular dependency detection"""
         self.check_circular_dependencies([])
 
         return self._async_start_recursive(network)
 
+    def network_ipv6(self, network: Network) -> str:
+        self._container.reload()
```
Comment (robertomonteromiguel) [inline utils/_context/containers.py:201]: why reload?

File: utils/_context/containers.py [hunk 6]
```diff
@@ -546,10 +570,9 @@ def __init__(self, host_log_folder, *, use_proxy=True, environment=None) -> None
             host_log_folder=host_log_folder,
             environment=environment,
             healthcheck={
-                "test": f"curl --fail --silent --show-error --max-time 2 http://localhost:{self.agent_port}/info",
+                "test": f"curl --fail --silent --show-error --max-time 2 http://localhost:{self.apm_receiver_port}/info",
```
Comment (robertomonteromiguel) [inline utils/_context/containers.py:573]: I guess for the healthcheck the "localhost" works, right? It's only for external access

**Review by robertomonteromiguel — COMMENTED** [review]
A thought: I imagined the IPV6 configuration as another parameter of the existing endtoend scenario or as an environment variable that makes everything work in IPV6.

File: utils/_context/containers.py (hunk 6 above)
Comment (cbeauchesne) [reply utils/_context/containers.py:573]: yes, because it is executed **inside** the container

File: utils/_context/containers.py (hunk 4 above)
Comment (cbeauchesne) [reply utils/_context/containers.py:190]: To be sure to have fresh data, see my comment above

File: docs/execute/IPv6.md (hunk 2 above)
Comment (cbeauchesne) [reply docs/execute/IPv6.md:1]: no, nowhere, what do you have in mind ?

File: .github/workflows/run-end-to-end.yml (hunk 1 above)
Comment (cbeauchesne) [reply .github/workflows/run-end-to-end.yml:120]: typoooo

File: docs/execute/IPv6.md (hunk 2 above)
Comment (robertomonteromiguel) [reply docs/execute/IPv6.md:1]: If not referenced from another page, this document does not explain much.
Or somewhere we explain the new scenario (docs/scenarios/ipv6.md).
Or from docs/execute/run.md, how to run scenarios over ipv6 and we reference this ipv6.md

File: utils/_context/_scenarios/endtoend.py (hunk 3 above)
Comment (cbeauchesne) [reply utils/_context/_scenarios/endtoend.py]: addressed in [3aae93a](https://github.com/DataDog/system-tests/pull/3624/commits/3aae93ab6692861a39a315dea348460a3119bb2b)

File: docs/execute/IPv6.md (hunk 2 above)
Comment (cbeauchesne) [reply docs/execute/IPv6.md]: addressed in [3aae93a](https://github.com/DataDog/system-tests/pull/3624/commits/3aae93ab6692861a39a315dea348460a3119bb2b)

File: .github/workflows/run-end-to-end.yml (hunk 1 above)
Comment (cbeauchesne) [reply .github/workflows/run-end-to-end.yml]: addressed in [3aae93a](https://github.com/DataDog/system-tests/pull/3624/commits/3aae93ab6692861a39a315dea348460a3119bb2b)

**Comment by cbeauchesne** [comment]
> A thought: I imagined the IPV6 configuration as another parameter of the existing endtoend scenario or as an environment variable that makes everything work in IPV6.

I understand the purpose, but it's not compatible as we don't use the proxy between the lib and the agent. And it would also add a dimensiion to the execution matrix and be a little bit an overkill.

**Review by robertomonteromiguel — APPROVED** [review]

**Pull request merged**